### DISCUSSION TITLE
Using vaxis with zig nightly builds?

### DISCUSSION THREAD

**Discussion by nurpax** [rockorager/libvaxis#138] [General]
Hi,

I normally develop with Zig nightly builds.  I wanted to try Vaxis but noticed that main is for zig 0.13.  OTOH, the vaxis zig-0.14.0 doesn't seem to build either due to errors like this:

```
C:Users\janne\AppData\Local\zig\p\1220dd654ef941fc76fd96f9ec6adadf83f69b9887a0d3f4ee5ac0a1a3e11be35cf5\src\ImageUnmanaged.zig:96:35: error: no field named 'Struct' in enum '@typeInfo(builtin.Type).@"union".tag_type.?'
            if (entryTypeInfo == .Struct) {
                                 ~^~~~~~
C:\Users\janne\scoop\apps\zig-dev\0.14.0-dev.2384\lib\std\builtin.zig:556:18: note: enum declared here
pub const Type = union(enum) {
                 ^~~~~
```

I guess Vaxis's "zig 0.14" was coded against some older Zig version than what I have?  I think the `.Struct` -> `.struct` was done a while ago.

I'm not sure how Zig projects generally handle these Zig versions.  Some projects tend to track "latest Zig I've used" in main and then branch off for Zig releases.

**Comment by rockorager** [comment]
Which zig nightly are you on? I am currently using `0.14.0-dev.2456+a68119f8f` on that branch and don't see this issue.

**Reply by nurpax** [reply]
Hmm, weird.  I'm on `0.14.0-dev.2384+cbc05e0b1` 

I tried building an example with `zig build example -Dexample=image`  but get an error like this:

```
C:\Users\janne\AppData\Local\zig\p\1220dd654ef941fc76fd96f9ec6adadf83f69b9887a0d3f4ee5ac0a1a3e11be35cf5\src\ImageUnmanaged.zig:96:35: error: no field named 'Struct' in enum '@typeInfo(builtin.Type).@"union".tag_type.?'
            if (entryTypeInfo == .Struct) {
                                 ~^~~~~~
C:\Users\janne\scoop\apps\zig-dev\0.14.0-dev.2384\lib\std\builtin.zig:556:18: note: enum declared here
pub const Type = union(enum) {
                 ^~~~~
```

Some other examples also failed to build.

Building another example `zig build example -Dexample=table`, I get this:

```
src\widgets\Table.zig:136:14: error: no field named 'Pointer' in enum '@typeInfo(builtin.Type).@"union".tag_type.?'
            .Pointer => |ptr| {
            ~^~~~~~~
C:\Users\janne\scoop\apps\zig-dev\0.14.0-dev.2384\lib\std\builtin.zig:556:18: note: enum declared here
pub const Type = union(enum) {
                 ^~~~~
referenced by:
    main: examples\table.zig:283:46
    WinStartup: C:\Users\janne\scoop\apps\zig-dev\0.14.0-dev.2384\lib\std\start.zig:617:37
```

I think the errors are due to the type enums being renamed to lowercase.

**Reply by nurpax** [reply]
The default example `text_input` builds though.

**Reply by rockorager** [reply]
Oh I see. So `image` is failing because I haven't updated zigimg yet.

`table` fails because I haven't updated that widget. I guess I better run through each example to make sure they work 😂

**Reply by nurpax** [reply]
Couple of other observations.  I'm running on Windows, in case that makes a difference.

- `nvim` example is listed in build help but fails to build due to `'examples\nvim.zig' failed: FileNotFound`
- `fuzzy` works but I get a segfault when I press enter and it exits:

```
C:\Users\janne\dev\libvaxis>zig build example -Dexample=fuzzy
Segmentation fault at address 0x162eac60000
C:\Users\janne\scoop\apps\zig-dev\0.14.0-dev.2384\lib\std\mem\Allocator.zig:314:26: 0x7ff702d23d24 in free__anon_16364 (example.exe.obj)
    @memset(non_const_ptr[0..bytes_len], undefined);
                         ^
C:\Users\janne\dev\libvaxis\src\Screen.zig:48:15: 0x7ff702cdb306 in deinit (example.exe.obj)
    alloc.free(self.buf);
              ^
C:\Users\janne\dev\libvaxis\src\Vaxis.zig:113:27: 0x7ff702cd1323 in deinit (example.exe.obj)
        self.screen.deinit(a);
                          ^
C:\Users\janne\dev\libvaxis\src\vxfw\App.zig:41:19: 0x7ff702cca248 in deinit (example.exe.obj)
    self.vx.deinit(self.allocator, self.tty.anyWriter());
                  ^
C:\Users\janne\dev\libvaxis\examples\fuzzy.zig:177:24: 0x7ff702cc9fc4 in main (example.exe.obj)
    errdefer app.deinit();
                       ^
C:\Users\janne\scoop\apps\zig-dev\0.14.0-dev.2384\lib\std\start.zig:436:53: 0x7ff702cccf37 in WinStartup (example.exe.obj)
    std.os.windows.ntdll.RtlExitUserProcess(callMain());
                                                    ^
???:?:?: 0x7ff836ec7373 in ??? (KERNEL32.DLL)
???:?:?: 0x7ff837dbcc90 in ??? (ntdll.dll)
example
└─ run example failure
error: the following command exited with error code 3:
```

**Reply by rockorager** [reply]
> * `nvim` example is listed in build help but fails to build due to `'examples\nvim.zig' failed: FileNotFound`

Well this is certainly a bug. I'll remove that. I used to have a neovim widget however I removed it because I didn't want to add something that had another dependency to the core library.

I *am* working on a neovim widget to work in the vxfw framework, however that will live in a separate repo.

> * `fuzzy` works but I get a segfault when I press enter and it exits:

This one is really strange. I don't get any such errors on linux...do you also get this segfault if you press ctrl+c to exit? (This one will exit with an error code, but it *shouldn't* segfault)

**Reply by nurpax** [reply]
I don't get a segfault if I ctrl-c, only when I hit enter.  Although I think I managed to get the fuzzy example to hang once when I was playing with it.. but that's harder to repro.

**Reply by nurpax** [reply]
BTW I can try to help debug this if you want help.  So feel free to suggest any debugging experiments :)

**Reply by rockorager** [reply]
Oh that would be helpful! I updated zigimg, so *that* one should compile, though I don't think it will work on Windows because of ConPTY.

The segfault: I'm guessing this happens because you don't have `fd` installed, so we get an error. The app tries to deinit, but it has an undefined buffer. Could you apply this diff, and see if you get a proper error with it?

```diff
diff --git a/examples/fuzzy.zig b/examples/fuzzy.zig
index 21d34ee30dfa..9089a29fd75e 100644
--- a/examples/fuzzy.zig
+++ b/examples/fuzzy.zig
@@ -174,7 +174,10 @@ pub fn main() !void {
     const allocator = gpa.allocator();
 
     var app = try vxfw.App.init(allocator);
-    errdefer app.deinit();
+    errdefer |err| {
+        std.log.err("error: {}", .{err});
+        app.deinit();
+    }
 
     const model = try allocator.create(Model);
     defer allocator.destroy(model);
diff --git a/src/Screen.zig b/src/Screen.zig
index ed2f3e1b70cd..ee7a780ffb6a 100644
--- a/src/Screen.zig
+++ b/src/Screen.zig
@@ -16,7 +16,7 @@ height: u16 = 0,
 width_pix: u16 = 0,
 height_pix: u16 = 0,
 
-buf: []Cell = undefined,
+buf: []Cell = &.{},
 
 cursor_row: u16 = 0,
 cursor_col: u16 = 0,
```

**Reply by nurpax** [reply]
In fact I do have `fd` installed.

I can see what the problem is:

```
    try app.run(model.widget(), .{});
    app.deinit();

    if (model.result.len > 0) {
        const writer = std.io.getStdOut().writer();
        try writer.print("{s}\n", .{model.result});
```

If the `writer.print` fails with an error, app is deinited twice because an `app.deinit()` was registered with errdefer after creating app.  Writing to stdout normally does work, even on Windows, but maybe it's the piping business before that messes it up somehow?

Had a better time building the samples now!  Thanks for fixing things!  

PS. If you know of any terminal emulators for Windows that'd support images, let me know. :)

**Reply by rockorager** [reply]
Ah good catch. Does it segfault if you move the deinit to after the write?

Windows Terminal supports sixels now, however libvaxis does not :)

**Reply by nurpax** [reply]
Yep, moving deinit after the write works.  The code with the errdefers is a little confusing but I guess it is what it is.

**Reply by rockorager** [reply]
All the examples should work in the zig-0.14.0 branch now

**Comment by rockorager** [comment]
> I'm not sure how Zig projects generally handle these Zig versions. Some projects tend to track "latest Zig I've used" in main and then branch off for Zig releases.

I've seen three approaches:

1. Always track `zig master`
2. `main`/`master` tracks `zig master`, and tag releases for zig versions
3. `main`/`master` tracks last zig tag. Update to `zig-next` after release

Personally, I prefer option 3. It means that I can add features or fix bugs to the library without forcing users to update zig as well. It's been working pretty well to have a longish running `zig-{next}` branch and rebasing on `main` periodically.
